src/instantiate.jl

```diff
@@ -0,0 +1,117 @@
+"""
+    struct OptimizerWithAttributes
+        optimizer
+        params::Vector{Pair{AbstractOptimizerAttribute,<:Any}}
+    end
+
+Object grouping an optimizer constructor and a list of optimizer attributes.
+Instances are created with [`instantiate`](@ref).
+"""
+struct OptimizerWithAttributes
+    # Function that takes zero arguments and returns a new optimizer.
+    # The type of the function could be
+    # * `Function`: a function, or
+    # * `DataType`: a type, or
+    # * `UnionAll`: a type with missing parameters.
+    optimizer
+    params::Vector{Pair{AbstractOptimizerAttribute,Any}}
+end
+
+_to_param(param::Pair{<:AbstractOptimizerAttribute}) = param
+_to_param(param::Pair{String}) = RawParameter(param.first) => param.second
+function _to_param(param::Pair)
+    error("Expected an optimizer attribute or a string, got `$(param.first)` which is a `$(typeof(param.first))`.")
+end
+function _add_params(params::Vector{Pair{AbstractOptimizerAttribute,Any}}) end
+function _add_params(params::Vector{Pair{AbstractOptimizerAttribute,Any}},
+                     param::Pair, other_params::Vararg{Pair, N}) where N
+    push!(params, _to_param(param))
+    _add_params(params, other_params...)
+end
+
+"""
+    OptimizerWithAttributes(optimizer_constructor, params::Pair...)
+
+Create an [`OptimizerWithAttributes`](@ref) with the parameters `params`.
+"""
+function OptimizerWithAttributes(optimizer_constructor, args::Vararg{Pair, N}) where N
+    if !applicable(optimizer_constructor)
+        error(_instantiate_not_callable_message)
+    end
+    params = Pair{AbstractOptimizerAttribute,Any}[]
+    _add_params(params, args...)
```

Comment:
Why not just:
```Julia
params = Pair{AbstractOptimizerAttribute,Any}[_to_param(arg) for arg in args]
```
Instead of all this `_add_params` indirection?

Reply:
The compiler might prefer `_add_params` recursion tuples of mixed type rather than for loops. In your approach, I would guess the dispatch to the `_to_param` method is done at runtime.

Reply:
This is very unlikely to be the bottleneck, and the current approach takes a little while to understand what's going on.